pkg/tsdb/azuremonitor/azuremonitor.go

```diff
@@ -30,56 +40,121 @@ func init() {
+		if err != nil {
+			return nil, err
+		}
+
+		jsonData := map[string]interface{}{}
+		err = json.Unmarshal(settings.JSONData, &jsonData)
+		if err != nil {
+			return nil, fmt.Errorf("error reading settings: %w", err)
+		}
+
+		model := datasourceInfo{}
+		err = json.Unmarshal(settings.JSONData, &model)
+		if err != nil {
+			return nil, fmt.Errorf("error reading settings: %w", err)
+		}
+		model.HTTPClient = client
+		model.URL = settings.URL
+		model.JSONData = jsonData
+		model.DecryptedSecureJSONData = settings.DecryptedSecureJSONData
+		model.DatasourceID = settings.ID
+
+		return model, nil
 	}
+}
 
+func newExecutor(im instancemgmt.InstanceManager, pm plugins.Manager, httpC httpclient.Provider, cfg *setting.Cfg) *AzureMonitorExecutor {
 	return &AzureMonitorExecutor{
-		httpClient:    httpClient,
-		dsInfo:        dsInfo,
-		pluginManager: s.PluginManager,
-		cfg:           s.Cfg,
-	}, nil
+		im:                 im,
+		httpClientProvider: httpC,
+		cfg:                cfg,
+		pluginManager:      pm,
+	}
+}
+
+func (s *Service) Init() error {
+	im := datasource.NewInstanceManager(NewInstanceSettings())
+	factory := coreplugin.New(backend.ServeOpts{
+		QueryDataHandler: newExecutor(im, s.PluginManager, s.HTTPClientProvider, s.Cfg),
+	})
+
+	if err := s.BackendPluginManager.Register(dsName, factory); err != nil {
+		azlog.Error("Failed to register plugin", "error", err)
+	}
+	return nil
 }
 
 // Query takes in the frontend queries, parses them into the query format
 // expected by chosen Azure Monitor service (Azure Monitor, App Insights etc.)
 // executes the queries against the API and parses the response into
 // the right format
-//nolint: staticcheck // plugins.DataPlugin deprecated
-func (e *AzureMonitorExecutor) DataQuery(ctx context.Context, dsInfo *models.DataSource,
-	tsdbQuery plugins.DataQuery) (plugins.DataResponse, error) {
+func (e *AzureMonitorExecutor) QueryData(ctx context.Context,
+	req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
 	var err error
 
-	var azureMonitorQueries []plugins.DataSubQuery
-	var applicationInsightsQueries []plugins.DataSubQuery
-	var azureLogAnalyticsQueries []plugins.DataSubQuery
-	var insightsAnalyticsQueries []plugins.DataSubQuery
-	var azureResourceGraphQueries []plugins.DataSubQuery
+	var azureMonitorQueries []backend.DataQuery
+	var applicationInsightsQueries []backend.DataQuery
+	var azureLogAnalyticsQueries []backend.DataQuery
+	var insightsAnalyticsQueries []backend.DataQuery
+	var azureResourceGraphQueries []backend.DataQuery
 
-	for _, query := range tsdbQuery.Queries {
-		queryType := query.Model.Get("queryType").MustString("")
+	i, err := e.im.Get(req.PluginContext)
+	if err != nil {
+		return nil, err
+	}
+
+	datasourceInfo := i.(datasourceInfo)
+	for _, query := range req.Queries {
+		model, err := simplejson.NewJson(query.JSON)
+		if err != nil {
+			return nil, err
+		}
+
+		queryType := model.Get("queryType").MustString("")
```

Comment:
Since you're supporting/using the `queryType` in model you can make use of the [query.QueryType](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go@v0.103.0/backend#DataQuery) and the `QueryTypeMux` which routes data queries to the correct query type handler, see example https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go@v0.103.0/backend/datasource#example-QueryTypeMux. Could be something to fix later to remove/clean up some code?

Reply:
I have given it a try, assuming there are no mixed query types, and simplify this code a bit. Let me know what you think.

Reply:
Yes it's possible to have a mix of query types. You can add one azure monitor query. Then add one more and choose azure log analytics for example

Reply:
:+1: 

Verified that this works as expected with a request with different types: 
```
curl 'http://localhost:3000/api/ds/query' ... \
  --data-raw '{"queries":[{"refId":"A","queryType":"Azure Log Analytics" ... \
                     {"refId":"B","queryType":"Azure Monitor"  ... \
                     {"refId":"C","queryType":"Azure Monitor"  ... \
```

The query handler receive requests with a single data type. With the example above, `mux.HandleFunc` gets executed once for the query `A` and again for `B, C`.  The result is properly joined as expected.

Reply:
:+1: 

Guess azure is a bit special since it has many different data sources baked into one and each one has their own struct. Your code for handling different queries could probably be simplified even more if each data source struct implements the [backend.QueryDataHandler](https://pkg.go.dev/github.com/grafana/grafana-plugin-sdk-go@v0.103.0/backend#QueryDataHandler) interface. However, this would require each data source struct to either embed the instance manager and/or something else to retrieve the dataSourceInfo/settings. Not crucial and something that could be worked on later, if needed.